### PR TITLE
CSV Output Plugin (115 redo)

### PR DESCRIPTION
**Pull request by williamzhang0306**
This branch contains only the commits related to the csv output plugin found in branch 115_csv_file_output. 

The plugin was modified so it checks and writes any high level data, not just 'learning_systems'. Although I'm not sure if that's how it's intended to be used.

There are also a few basic unit tests for the csv output plugin.

### CODE REVIEW

File: plugins/csv_output/csv_output_plugin.py [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+# Copyright © 2023 United States Government as represented by the Administrator of
+# the National Aeronautics and Space Administration. No copyright is claimed in the
+# United States under Title 17, U.S. Code. All Other Rights Reserved.
+#
+# Licensed under the NASA Open Source Agreement version 1.3
+# See "NOSA GSC-19165-1 OnAIR.pdf"
+
+from datetime import datetime
+
+#import numpy as np
```
Comment (the-other-james) [inline plugins/csv_output/csv_output_plugin.py:12]: Remove commented out import

File: plugins/csv_output/csv_output_plugin.py [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+        self.lines_current = 0
+        self.current_buffer = [] # List of telemetry points
+        self.filename_preamble = "csv_out_"
+        self.filename = ""
+
+        # Plugin should write each frame of data out to a .csv file
+        # low_level_data is optional. Should use the headers provided in __init__
+        # high level data is coming from different sources: list in headers
+
+        # simplest mode is just always writing out to a file
```
Comment (the-other-james) [inline plugins/csv_output/csv_output_plugin.py:31]: This comment is out of date

File: plugins/csv_output/csv_output_plugin.py [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+                for telem_point in plugin_output:
+                    self.current_buffer.append(str(telem_point))
+                    
+    def render_reasoning(self):
+        """
+        System should return its diagnosis
+        """
+
+        if (self.first_frame):
+            # Create file- TODO: make separate function
```
Comment (the-other-james) [inline plugins/csv_output/csv_output_plugin.py:66]: Comment is no longer accurate

File: plugins/csv_output/csv_output_plugin.py [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+            date_stamp = datetime.today().strftime('%j_%H_%M')
+            self.file_name = self.filename_preamble + date_stamp + ".csv"
+
+            # Write out to file
```
Comment (the-other-james) [inline plugins/csv_output/csv_output_plugin.py:70]: Write out headers to file

File: plugins/csv_output/csv_output_plugin.py [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+            with open(self.file_name, 'a') as file:
+                delimiter = ','
+                file.write(delimiter.join(self.headers) + '\n')
+            self.first_frame = False
+        pass
+
+        # Write out to file
```
Comment (the-other-james) [inline plugins/csv_output/csv_output_plugin.py:77]: Write out data frame to file

File: test/plugins/csv_output/test_csv_output_plugin.py [hunk 6]
```diff
@@ -0,0 +1,239 @@
...
+
+
+def test_init_initalizes_expected_default_variables():
+
+    arg_name = MagicMock()
+    arg_headers = [MagicMock(), MagicMock()]
+
+    csv_out = CSV_Output(arg_name, arg_headers)
+
+    # could we not just csv_out = CSV_Output(arg_name, arg_headers)
```
Comment (the-other-james) [inline test/plugins/csv_output/test_csv_output_plugin.py:27]: Remove comment?

File: test/plugins/csv_output/test_csv_output_plugin.py [hunk 7]
```diff
@@ -0,0 +1,239 @@
...
+    csv_out.render_reasoning()
+
+    # check if a new file was set, but NOT yet created
+    try:
+        assert csv_out.file_name != old_file_name
+        assert not os.path.exists(csv_out.file_name)
+    except:
+        raise AssertionError("file name was not updated or new file was created prematurely")
+    finally:
+        os.remove(old_file_name)
```
Comment (the-other-james) [inline test/plugins/csv_output/test_csv_output_plugin.py:239]: Small nitpick: please end files with a newline. That way when someone cats the file in a terminal it looks nicer.

**Review by the-other-james — CHANGES_REQUESTED** [review]
Looks good! Most of my comments are just to bring the code comments into line with what the code is doing (which is probably my fault anyways).

Test coverage is 100% and I confirmed it runs on my machine. One bit of unexpected behavior is that the output files are larger than 10 lines because OnAIR is running so fast that "new" filename is the same as the old. I don't think this needs to be "fixed" since no data is lost... but a comment to this effect would be good to have next to where lines_per_file is set.

**Comment by williamzhang0306** [comment]
Thanks! I'll start getting those changes in

**Review by the-other-james — APPROVED** [review]
Looks good, thanks!

**Pull request merged**